pkg/controller/csi_handler.go

```diff
@@ -253,14 +271,20 @@ func (h *csiHandler) csiAttach(va *storage.VolumeAttachment) (*storage.VolumeAtt
 		return va, nil, err
 	}
 
-	nodeID, err := h.getNodeID(h.attacherName, va.Spec.NodeName)
+	nodeID, err := h.getNodeID(h.attacherName, va.Spec.NodeName, nil)
 	if err != nil {
 		return va, nil, err
 	}
 
-	va, err = h.addVAFinalizer(va)
-	if err != nil {
-		return va, nil, fmt.Errorf("could not add VolumeAttachment finalizer: %s", err)
+	originalVA := va
+	va, finalizerAdded := h.prepareVAFinalizer(va)
+	va, nodeIDAdded := h.prepareVANodeID(va, nodeID)
+	if finalizerAdded || nodeIDAdded {
+		va, err = h.saveVA(va)
+		if err != nil {
+			// va modification failed, return the original va that's still on API server
```

Comment:
is this necessarily true, could we have error type where API Server has been modified yet error isi still returned. How bad is returning the incorrect VA here (it might not be that important).

Reply:
The code tries to save error to `va.Status`. It's likely fail if the error was conflict, but who knows... I try to keep `va` to the last state acknowledged by API server.